dali/kernels/imgproc/color_manipulation/color_space_conversion_impl.h

```diff
@@ -39,172 +39,263 @@ DALI_HOST_DEV DALI_FORCEINLINE vec<N, float> norm(vec<N, float> x) {
-  constexpr vec3 coeffs(0.257f, 0.504f, 0.098f);
-  auto rgb = detail::norm(rgb_in);  // TODO(janton): optimize number of multiplications
-  return ConvertSatNorm<Output>(dot(coeffs, rgb) + 0.0625f);
+template <typename Input>
+DALI_HOST_DEV DALI_FORCEINLINE float norm(Input x) {
+  return ConvertNorm<float>(x);
 }
 
-template <>
-DALI_HOST_DEV DALI_FORCEINLINE uint8_t rgb_to_y(vec<3, uint8_t> rgb) {
-  constexpr vec3 coeffs(0.257f, 0.504f, 0.098f);
-  return ConvertSat<uint8_t>(dot(coeffs, rgb) + 16.0f);
+template <typename From, typename To>
+constexpr DALI_HOST_DEV float scale_factor() {
+  constexpr double to = is_fp_or_half<To>::value
+                      ? 1.0 : static_cast<double>(max_value<To>());
+
+  constexpr double from = is_fp_or_half<From>::value
+                        ? 1.0 : static_cast<double>(max_value<From>());
+
+  constexpr float factor = to / from;
+  return factor;
 }
 
-template <typename Output, typename Input>
-DALI_HOST_DEV DALI_FORCEINLINE Output rgb_to_cb(vec<3, Input> rgb_in) {
-  constexpr vec3 coeffs(-0.148f, -0.291f, 0.439f);
-  auto rgb = detail::norm(rgb_in);  // TODO(janton): optimize number of multiplications
-  return ConvertSatNorm<Output>(dot(coeffs, rgb) + 0.5f);
+template <typename T>
+constexpr DALI_HOST_DEV std::enable_if_t<std::is_integral<T>::value, float> bias_scale() {
+  // The scale is 2^positive_bits
+  // Since that value always overflows, we shift by bits-1 and then multiply x2 already in
+  // floating point.
+  return 2.0f * static_cast<float>(T(1) << (sizeof(T) * 8 - 1 - std::is_signed<T>::value));
 }
 
-template <>
-DALI_HOST_DEV DALI_FORCEINLINE uint8_t rgb_to_cb(vec<3, uint8_t> rgb) {
-  constexpr vec3 coeffs(-0.148f, -0.291f, 0.439f);
-  if (rgb.x == rgb.y && rgb.x == rgb.z) return 128;
-  return ConvertSat<uint8_t>(dot(coeffs, rgb) + 128.0f);
+template <typename T>
+constexpr DALI_HOST_DEV std::enable_if_t<!std::is_integral<T>::value, float> bias_scale() {
+  return 1.0f;
 }
 
-template <typename Output, typename Input>
-DALI_HOST_DEV DALI_FORCEINLINE Output rgb_to_cr(vec<3, Input> rgb_in) {
-  constexpr vec3 coeffs(0.439f, -0.368f, -0.071f);
-  auto rgb = detail::norm(rgb_in);  // TODO(janton): optimize number of multiplications
-  return ConvertSatNorm<Output>(dot(coeffs, rgb) + 0.5f);
+}  // namespace detail
+
+// Y, Cb, Cr definition from ITU-R BT.601, with values in the range 16-235, allowing for
+// footroom and headroom
+struct itu_r_bt_601 {
+  template <typename Output, typename Input>
+  static DALI_HOST_DEV DALI_FORCEINLINE constexpr Output rgb_to_y(vec<3, Input> rgb) {
+    constexpr vec3 coeffs = vec3(0.25678823529f, 0.50412941176f, 0.09790588235f)
+                            * detail::scale_factor<Input, Output>();
+    constexpr float bias = 0.0625f * detail::bias_scale<Output>();
+    float y = dot(coeffs, rgb) + bias;
+    return needs_clamp<Input, Output>::value ? Convert<Output>(y) : ConvertSat<Output>(y);
```

Comment:
It would be a very weird function, because it doesn't convert from Input, but from an unnormalized float. - normally, ConvertSat doesn't clamp when it doesn't have to. Here we know that the intermediate float value is in range.